### PR TITLE
CSHARP-2796: Add API for specifying a custom endpoint with AWS masterkey provider.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
Spec: https://jira.mongodb.org/browse/SPEC-1431

Evergreen: https://evergreen.mongodb.com/version/5db06d0c57e85a7eedea9d5c (Hopefully I generated this link correctly)

### CODE REVIEW

File: bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs [hunk 1]
```diff
@@ -260,6 +260,50 @@ public void EncryptExplicitStepwise()
...
+                customerMasterKey: "cmk",
+                region: "us-east-1",
+                endpoint: endpoint);
+            var key = new AwsKmsCredentials(awsSecretAccessKey: "us-east-1", awsAccessKeyId: "us-east-1");
+
+            using (var cryptClient = CryptClientFactory.Create(new CryptOptions(CreateCredentialsMap(key))))
+            using (var context = cryptClient.StartCreateDataKeyContext(keyId))
+            {
+                var (_, dataKeyDocument) = ProcessContextToCompletion(context, isKmsDecrypt: false);
+                dataKeyDocument.Should().NotBeNull();
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs]: Vincent,  `dataKeyDocument` doesn't contain the `endpoint` value. Do we expect this value here?

File: bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs [hunk 2]
```diff
@@ -260,6 +260,50 @@ public void EncryptExplicitStepwise()
...
+            using (var cryptClient = CryptClientFactory.Create(new CryptOptions(CreateCredentialsMap(key))))
+            using (var context = cryptClient.StartCreateDataKeyContext(keyId))
+            {
+                BsonDocument dataKeyDocument;
+                var (state, _, _) = ProcessState(context, isKmsDecrypt: false);
+                state.Should().Be(CryptContext.StateCode.MONGOCRYPT_CTX_NEED_KMS);
+
+                (state, _, dataKeyDocument) = ProcessState(context, isKmsDecrypt: false);
+                state.Should().Be(CryptContext.StateCode.MONGOCRYPT_CTX_READY);
+                dataKeyDocument.Should().NotBeNull();
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs]: The same question as above.

File: bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs [hunk 3]
```diff
@@ -32,10 +31,27 @@ public class AwsKeyId : IKmsKeyId, IInternalKmsKeyId
         /// <param name="customerMasterKey">The Amazon Resource Name (ARN) of the customer master key.</param>
         /// <param name="region">The region.</param>
         public AwsKeyId(string customerMasterKey, string region)
+            : this(
```
Comment (rstam) [inline bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs]: Why so many overloads? Every time we add a new optional parameter this technique doubles the number of overloads.

Why not a single constructor with optional parameters?

File: bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs [hunk 4]
```diff
@@ -78,24 +121,41 @@ void IInternalKmsKeyId.SetCredentials(ContextSafeHandle context, Status status)
...
+                        status,
+                        Library.mongocrypt_ctx_setopt_masterkey_aws_endpoint(context, endPointKeyPointer, -1));
+                }
+                finally
+                {
+                    Marshal.FreeHGlobal(endPointKeyPointer);
+                }
+            }
+
+            ((IInternalKmsKeyId)this).SetAlternateKeyNames(context, status);
```
Comment (rstam) [inline bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs:115]: I assume moving this here (outside of the try/finally block it used to be in) is safe?

**Review by rstam — APPROVED** [review]
Other than a comment on the doubling of the number of constructor overloads and a suggestion to consider optional parameters instead LGTM.

File: bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs:115]: yes. `try/finally` is related to `mongocrypt_ctx_setopt_masterkey_aws`. So, no reasons also to have `SetAlternateKeyNames` inside this block too

File: bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs]: Fixed.

File: bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs [hunk 5]
```diff
@@ -46,11 +32,17 @@ public AwsKeyId(string customerMasterKey, string region)
...
+            string customerMasterKey,
+            string region,
+            IEnumerable<byte[]> alternateKeyNamesBsonDocuments = null,
+            string endpoint = null)
         {
-            Region = region;
             CustomerMasterKey = customerMasterKey;
-            AlternateKeyNameBsonDocuments = alternateKeyNamesBsonDocuments.ToList().AsReadOnly();
+            Region = region;
+            AlternateKeyNameBsonDocuments = (alternateKeyNamesBsonDocuments ?? Enumerable.Empty<byte[]>()).ToList();
```
Comment (vincentkam) [inline bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs]: I believe by adding`AsReadOnly`, a caller will be unable to cast `AlternateKeyNameBsonDocuments` to a be a mutable list.
`AlternateKeyNameBsonDocuments = (alternateKeyNamesBsonDocuments ?? Enumerable.Empty<byte[]>()).ToList().AsReadOnly();`

File: bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs [hunk 6]
```diff
@@ -78,24 +78,41 @@ void IInternalKmsKeyId.SetCredentials(ContextSafeHandle context, Status status)
...
+
+            if (Endpoint != null)
+            {
+                IntPtr endPointKeyPointer = (IntPtr)Marshal.StringToHGlobalAnsi(Endpoint);
+                try
+                {
+                    // Let mongocrypt run strlen
+                    context.Check(
+                        status,
+                        Library.mongocrypt_ctx_setopt_masterkey_aws_endpoint(context, endPointKeyPointer, -1));
```
Comment (vincentkam) [inline bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs:107]: -1 is fine here, after some discussion with Kevin :)

File: bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs (hunk 1 above)
Comment (vincentkam) [reply bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs]: Spoke with Kevin, there may be a bug in libmongocrypt

Comment (vincentkam) [reply bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs]: Let's change the test to expect the endpoint value in anticipation of the new version of libmongocrypt :)

File: bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs (hunk 2 above)
Comment (vincentkam) [reply bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs]: Same answer as above 😸

**Review by vincentkam — CHANGES_REQUESTED** [review]

File: bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/AwsKeyId.cs]: Not 100% that we need to worry about it. However, I think there is no harm to add it too

File: bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs]: done

File: bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs]: done

**Review by DmitryLukyanov — COMMENTED** [review]
Evergreen:  https://evergreen.mongodb.com/version/5db1e6ec3627e01a3b53b948
Ready to review.

**Review by vincentkam — APPROVED** [review]
LGTM! 😸

**Pull request closed without merging**